test/project_types/script/ui/error_handler_test.rb

```diff
@@ -18,16 +18,35 @@
     end
 
     describe "when failed operation message, cause of error, and help suggestion are all provided" do
-      it "should abort with the cause of error and help suggestion" do
-        if ci?
-          $stderr.expects(:puts).with("✗ Error")
-          $stderr.expects(:puts).with("#{failed_op} #{cause_of_error} #{help_suggestion}")
-        else
-          $stderr.expects(:puts).with("\e[0;31m✗ Error\e[0m")
-          $stderr.expects(:puts).with("\e[0m#{failed_op} #{cause_of_error} #{help_suggestion}")
-        end
-        assert_raises(ShopifyCLI::AbortSilent) do
-          subject
+      describe "when failed operation message and cause of error end with whitespace" do
+        let(:failed_op) { "Operation didn't complete. " }
+        let(:cause_of_error) { "This is why it failed.\n" }
+        it "should abort with the cause of error and help suggestion" do
+          if ci?
+            $stderr.expects(:puts).with("✗ Error")
+            $stderr.expects(:puts).with("#{failed_op}#{cause_of_error}#{help_suggestion}")
+          else
+            $stderr.expects(:puts).with("\e[0;31m✗ Error\e[0m")
+            $stderr.expects(:puts).with("\e[0m#{failed_op}#{cause_of_error}#{help_suggestion}")
+          end
+          assert_raises(ShopifyCLI::AbortSilent) do
+            subject
+          end
+        end
+      end
+
+      describe "when failed operation message and cause of error do not end with whitespace" do
+        it "should abort with the cause of error and help suggestion" do
+          if ci?
+            $stderr.expects(:puts).with("✗ Error")
+            $stderr.expects(:puts).with("#{failed_op} #{cause_of_error} #{help_suggestion}")
+          else
+            $stderr.expects(:puts).with("\e[0;31m✗ Error\e[0m")
+            $stderr.expects(:puts).with("\e[0m#{failed_op} #{cause_of_error} #{help_suggestion}")
+          end
+          assert_raises(ShopifyCLI::AbortSilent) do
+            subject
+          end
```

Comment:
These two seem identical. I get that in one of them we've got trailing whitespace but that doesn't affect the outcome, so it looks to me like we're just running the same test twice? (or put another way, we're testing the private method, which we shouldn't need to do).

Reply:
We want to test that when there is trailing whitespace, we don't add in a space, but when there isn't trailing whitespace, we do add a space. There is a case without trailing whitespace where we test that the space is added. There is a case with trailing whitespace where we test that the space is not added. I'm not sure how else to do it.

Reply:
@lopert: good call out!  👍 
My 2 cents: the unit test here doesn't need to actually call `CLI::UI.fmt`.  That should be `CLI::UI.fmt.expects(...).with(...).returns(msg)`, and then, `stderr.expects(:puts).with(msg)` is sufficient.

Reply:
I'm a bit confused. @ErinRenNumber1 are you saying we should change the test or are the expectations on `$stderr` sufficient?

Reply:
I was just re-using the existing test structure but adding an extra case (not related to CI, that was a pattern that was already there). It appears that when we run the tests locally, we need to include the escape sequences for colouring the output (`\e[0;31m` to make the output red, and `\e[0m` to reset it; see more on the escape sequences [here](https://en.wikipedia.org/wiki/ANSI_escape_code#3-bit_and_4-bit)), whereas in CI we don't include those characters. My guess is that it is a quirk of GitHub actions, but @jacobsteves can confirm as he added this code originally. Other than those colour-related escape sequences, the assertions in both branches of the if-else should be identical.

Reply:
Unfortunately, I can't remember why I added that `ci?` check but your guess of a GitHub actions quirk seems reasonable. I agree that it's weird behaviour. We could also consider using `io = capture_io { subject }` and validate the output using `assert_message_output` or `assert_matches` which should strip the escape sequences and allow us to avoid the `ci?` check.

Reply:
I did two implementations, the first just uses `capture_io` and a regex to remove ANSI colour escape sequences but not `assert_message_output`. I believe it is a more thorough test because it asserts on the full content of the error output, including order, rather than what `assert_message_output` does (only checks that each of the input strings is present somewhere in the output, doesn't care about order). The second uses `capture_io` and `assert_message_output`, so it conforms more to the pattern of the codebase. Let me know what you think is better.

Reply:
Nice! Thanks for that, the commit history is very helpful. I think both approaches make sense! I do prefer your first approach since I agree that it's more of a thorough test. I added [one comment](https://github.com/Shopify/shopify-cli/pull/1996#discussion_r797586877) which I think would generalize things a bit more so its easier to use in other places as well.

But if you want to stick with the latter approach, I'm fine with that too since thats consistent with the rest of the codebase 🤷‍♂️

Reply:
With your suggestion for using `CLI::UI::ANSI.strip_codes`, I really like the first approach. Thanks Jacob!

Reply:
@lopert @ErinRenNumber1 does this change address your concerns?

Reply:
Yes, looks way cleaner now! 👍